contracts/BSCValidatorSet.sol

```diff
@@ -479,7 +501,178 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
+      return index;
+    }
+    // the actually index
+    index = index - 1;
+
+    uint256 income = currentValidatorSet[index].incoming;
+    currentValidatorSet[index].incoming = 0;
+    uint256 rest = currentValidatorSet.length - 1;
+    emit validatorMisdemeanor(validator,income);
+    if (rest==0) {
+      // should not happen, but still protect
+      return index;
+    }
+    uint256 averageDistribute = income/rest;
+    if (averageDistribute!=0) {
+      for (uint i=0;i<index;i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+      uint n = currentValidatorSet.length;
+      for (uint i=index+1;i<n;i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    return index;
+  }
+
+  function _felony(address validator) private {
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
+      return;
+    }
+    // the actually index
+    index = index - 1;
+    uint256 income = currentValidatorSet[index].incoming;
+    uint256 rest = currentValidatorSet.length - 1;
+    if (getValidators().length <= 1) {
+      // will not remove the validator if it is the only one validator.
+      currentValidatorSet[index].incoming = 0;
+      return;
+    }
+    emit validatorFelony(validator,income);
+
+    _removeFromCurrentValidatorSet(validator, index);
+
+    uint256 averageDistribute = income/rest;
+    if (averageDistribute!=0) {
+      uint n = currentValidatorSet.length;
+      for (uint i=0;i<n;i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+  }
+
+  function _removeFromCurrentValidatorSet(address validator, uint256 index) private {
+    delete currentValidatorSetMap[validator];
+    // It is ok that the validatorSet is not in order.
+    if (index != currentValidatorSet.length - 1) {
+      currentValidatorSet[index] = currentValidatorSet[currentValidatorSet.length - 1];
+      currentValidatorSetMap[currentValidatorSet[index].consensusAddress] = index + 1;
+    }
+    currentValidatorSet.pop();
+  }
+
+  function _forceMaintainingValidatorsExit(Validator[] memory validatorSet) private returns (Validator[] memory unjailedValidatorSet){
+    uint256 numOfFelony = 0;
+    address validator;
+    bool isFelony;
+
+    // 1. validators exit maintenance
+    for (uint i = 0; i < maintainingValidatorSet.length; i++) {
+      validator = maintainingValidatorSet[i].consensusAddress;
+      // exit maintenance
+      isFelony = _exitMaintenance(validator);
```

Comment:
No need to `_exitMaintenance` if it is not `isMaintaining`

Reply:
each validator in `maintainingValidatorSet ` must be `isMaintaining`